SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp

```diff
@@ -299,114 +162,118 @@ void USpatialWorkerConnection::ConnectToLocator()
 
 	Worker_ConnectionFuture* ConnectionFuture = Worker_Alpha_Locator_ConnectAsync(WorkerLocator, &ConnectionParams);
 
+	FinishConnecting(ConnectionFuture);
+}
+
+void USpatialWorkerConnection::FinishConnecting(Worker_ConnectionFuture* ConnectionFuture)
+{
 	AsyncTask(ENamedThreads::AnyBackgroundThreadNormalTask, [ConnectionFuture, this]
 	{
-		WorkerConnection = Worker_ConnectionFuture_Get(ConnectionFuture, nullptr);
-
+		Worker_Connection* NewWorkerConnection = Worker_ConnectionFuture_Get(ConnectionFuture, nullptr);
 		Worker_ConnectionFuture_Destroy(ConnectionFuture);
-		if (Worker_Connection_IsConnected(WorkerConnection))
+
+		AsyncTask(ENamedThreads::GameThread, [this, NewWorkerConnection]
 		{
+			this->WorkerConnection = NewWorkerConnection;
+
 			CacheWorkerAttributes();
-			AsyncTask(ENamedThreads::GameThread, [this]
+
+			if (Worker_Connection_IsConnected(NewWorkerConnection))
 			{
-				this->OnConnectionSuccess();
-			});
-		}
-		else
-		{
-			// TODO: Try to reconnect - UNR-576
-			AsyncTask(ENamedThreads::GameThread, [this]
+				OnConnectionSuccess();
+			}
+			else
 			{
-				this->OnConnectionFailure();
-			});
-		}
+				// TODO: Try to reconnect - UNR-576
+				OnConnectionFailure();
+			}
+		});
 	});
 }
 
 SpatialConnectionType USpatialWorkerConnection::GetConnectionType() const
 {
-	// The legacy locator path did not specify PlayerIdentityToken, so if we have one
-	// we can use the new locator workflow.
 	if (!LocatorConfig.PlayerIdentityToken.IsEmpty())
 	{
 		return SpatialConnectionType::Locator;
 	}
-	else if (!LegacyLocatorConfig.LoginToken.IsEmpty())
-	{
-		return SpatialConnectionType::LegacyLocator;
-	}
 	else
 	{
 		return SpatialConnectionType::Receptionist;
 	}
 }
 
-Worker_OpList* USpatialWorkerConnection::GetOpList()
+TArray<Worker_OpList*> USpatialWorkerConnection::GetOpList()
 {
-	return Worker_Connection_GetOpList(WorkerConnection, 0);
+	TArray<Worker_OpList*> OpLists;
+	while (!OpListQueue.IsEmpty())
+	{
+		Worker_OpList* OutOpList;
+		OpListQueue.Dequeue(OutOpList);
+		OpLists.Add(OutOpList);
+	}
+
+	return OpLists;
 }
 
 Worker_RequestId USpatialWorkerConnection::SendReserveEntityIdsRequest(uint32_t NumOfEntities)
 {
-	return Worker_Connection_SendReserveEntityIdsRequest(WorkerConnection, NumOfEntities, nullptr);
+	QueueOutgoingMessage<FReserveEntityIdsRequest>(NumOfEntities);
+	return NextRequestId++;
```

Comment:
Is it safe to rely that these will match request IDs returned by C API?

Reply:
I looked at the worker SDK code and I'm basically doing the same thing. Found a thread as well in worker support which said request ids are unique per connection.

Reply:
so I think it is safe.

Reply:
Unique per connection doesn't mean monotonic increasing, although it's the most obvious thing to do. I'm pretty sure this will work for now, but ultimately we should probably handle these in a way that means we don't have to guess. Most likely the command book keeping is written to at the point we send stuff. We don't actually need to export these Ids to users to we can get away without having too much state.